Feat/validator: Create new Sudoku class in sudoku.py and make a small change to grid.py for easier Square attributes handling

alecpramanik opened the pull request:
## This closes #3.

## New Feature: Sudoku Class and Validation

A new `Sudoku` class has been created in `sudoku.py` which contains methods for inserting and removing numbers in the `Sudoku` grid, as well as checking if a number is valid to be inserted in a specified row and column of the Sudoku grid according to the rules of sudoku. Additionally, a `sudoku_demo.py` file was added for manual testing.

This implementation takes inspiration from some of the work done on branch https://github.com/BevapnadeBerit/Proj_INDA/tree/feat/puzzle_maker: 9b1f7ddfdd55f6ebfea92fb9e47ae74c67d8e049.
Functionality necessary for a backtracking algorithm, such as finding empty squares, will be implemented in subsequent features.

## IMPORTANT: Changes to `grid.py`, only merge if acceptable
Some small changes have been made to the `Grid` class in `grid.py` to more conveniently implement `Sudoku`. 

- The `Square` class now also stores the additional fields `row` and `col` for easier access when checking for values in `Squares`. 
- The `Box` class in grid.py has also been modified to pass the new `row` and `col` values of each square to the `Square` object upon initialization. 
- `Square's` `set_value` method has been modified to set `self.number` to `None` if the new value is `-1`. This is so that `Sudoku's` `remove_number` method properly resets  `Square's` `number` field, not just the `value`.

See changes to `grid.py` for the specifics.

BevapnadeBerit reviewed (commented):
In grid.py row 54-55 you initialize variables row and col but I don't see them used, have I missed something? Did you forget to put them into the Square constructor on row 57?

Otherwise, everything looks good.

alecpramanik commented on grid.py:
```diff
@@ -27,19 +27,20 @@ def __init__(self, grid_pos: tuple[int, int], v_boxes: int, h_boxes: int, box_wi
 
         self.boxes: [[Box]] = []
-        for h in range(h_boxes):
+        for v in range(v_boxes):
```
Flipped v_boxes and h_boxes so that the Boxes can be intuitively accessed using squares[box_row, box_col].

alecpramanik commented on grid.py:
```diff
@@ -48,26 +49,30 @@ def __init__(self, box_pos: tuple[int, int], width: int, height: int, group: pyg
 
         self.squares: [[Square]] = []
-        for x in range(width):
+        for y in range(height):
```
Flipped height and width in the nested loop so that the Squares can be intuitively accessed using squares[row % 3, col % 3].

alecpramanik commented:
> In grid.py row 54-55 you initialize variables row and col but I don't see them used, have I missed something? Did you forget to put them into the Square constructor on row 57?
> 
> Otherwise, everything looks good.

Nice catch! I didn't notice it because `sudoku_demo.py` was only populating the first Box (0, 0) and `_get_box` was still giving the correct coordinates for that box with `square.col // 3` == 0 and `square.row // 3` == 0.

I have made the following major changes:
- Updated the `Grid` class to assign `box_row` and `box_col` values to each `Box`, which places them in a 3x3 coordinate which can then be used to find `Squares` by using `grid.boxes[row // 3][col // 3].squares[row % 3][col % 3]`, where `row` and `col` are 9x9 coordinates. 
- In order to make this work, I had to flip the nested loops mentioned in the comments above. 
- The `Box` class should now be correctly calculating and assigning `row` and `col` values to `Squares`.
- Made the `Sudoku` class more modular with additional private methods.
- Updated `sudoku_demo.py` with a more thorough demonstration.

BevapnadeBerit reviewed (approved):
Looks good
